pkg/repro/repro.go

```diff
@@ -603,126 +605,38 @@ func (ctx *context) returnInstance(inst *reproInstance) {
 }
 
 func (ctx *context) reproLogf(level int, format string, args ...interface{}) {
+	if ctx.logf != nil {
+		ctx.logf(format, args...)
+	}
 	prefix := fmt.Sprintf("reproducing crash '%v': ", ctx.crashTitle)
 	log.Logf(level, prefix+format, args...)
 	ctx.stats.Log = append(ctx.stats.Log, []byte(fmt.Sprintf(format, args...)+"\n")...)
 }
 
 func (ctx *context) bisectProgs(progs []*prog.LogEntry, pred func([]*prog.LogEntry) (bool, error)) (
 	[]*prog.LogEntry, error) {
+	// Set up progs bisection.
 	ctx.reproLogf(3, "bisect: bisecting %d programs", len(progs))
-
-	ctx.reproLogf(3, "bisect: executing all %d programs", len(progs))
-	crashed, err := pred(progs)
-	if err != nil {
-		return nil, err
-	}
-	if !crashed {
-		ctx.reproLogf(3, "bisect: didn't crash")
-		return nil, nil
-	}
-
-	guilty := [][]*prog.LogEntry{progs}
-again:
-	if len(guilty) > 8 {
-		// This is usually the case for flaky crashes. Continuing bisection at this
-		// point would just take a lot of time and likely produce no result.
+	ret, err := generic.MinimizeSlice(generic.MinimizeConfig[*prog.LogEntry]{
+		Pred: func(progs []*prog.LogEntry) (bool, error) {
+			// Don't waste time testing empty crash log.
+			if len(progs) == 0 {
+				return false, nil
+			}
+			return pred(progs)
+		},
+		// For flaky crashes we usually end up with too many chunks.
+		// Continuing bisection would just take a lot of time and likely produce no result.
+		MaxChunks: 8,
+		Logf: func(msg string, args ...interface{}) {
+			ctx.reproLogf(3, "bisect: "+msg, args...)
+		},
+	}, progs)
+	if err == generic.ErrTooManyChunks {
```

Comment:
Do we really want to abort in this case? The original comment says "Continuing bisection at this point would just take a lot of time", and we already have MaxSteps for this. Perhaps we just set some reasonable MaxSteps and simplify logic?

Reply:
I'd say MaxSteps means "let's do as much as we can in the limited number of steps and continue with what we got", which is not what we really want for reproducer minimization. Further bisection steps are not a waste of time in this case.

But the limit on the chunks number can be seen as a kind of result sanitization. If we already have 8 chunks, then we already know we need at least 8 programs from the log, possibly even more when once the algorithm finishes. Should that ever be happening? Won't we go crazy glueing, running and minimizing so many calls later?